### PR TITLE
Propagate Ledger Tables to Cardano test packages

### PR DESCRIPTION
**Pull request by Jasagredo**
# Description

This PR contains the changes required to propagate the Ledger Tables to the test packages of the Cardano bundle.

# Checklist

- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [x] The documentation has been properly updated
    - [x] New tests are added if needed and existing tests are updated
    - [ ] Any changes affecting Consensus packages must have an entry in the appropriate `changelog.d` directory created using [`scriv`](https://github.com/input-output-hk/scriv). If in doubt, see the [Consensus release process](../ouroboros-consensus/docs/ReleaseProcess.md).
    - [ ] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [x] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [x] Reviewer requested

### CODE REVIEW

**Comment by Jasagredo** [comment]
Current dependencies on/for this PR:
* master
  * **PR #4352** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4352" target="_blank">Graphite</a> 
    * **PR #4341** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4341" target="_blank">Graphite</a> 
      * **PR #4342** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4342" target="_blank">Graphite</a> 
        * **PR #4343** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4343" target="_blank">Graphite</a> 
    * **PR #4316** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4316" target="_blank">Graphite</a> 
    * **PR #4344** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4344" target="_blank">Graphite</a> 
      * **PR #4366** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4366" target="_blank">Graphite</a> 
        * **PR #4363** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4363" target="_blank">Graphite</a> 
          * **PR #4391** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4391" target="_blank">Graphite</a> 
            * **PR #4417** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4417" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4417?utm_source=stack-comment).

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+type Crypto = MockCryptoCompatByron
+type Proto  = TPraos Crypto
+
+tests :: TestTree
+tests = testGroup "UpdateTablesOnEraTransition"
+  [ testTablesTranslation "Byron to Shelley"
+                          byronToShelleyLedgerStateTranslation
+                          byronUtxosAreInsertsInShelleyUtxoDiff
+                          (\st -> cover 50 (      nonEmptyUtxosByron st) "UTxO set is not empty"
+                                -- FIXME: we should test with empyt UTxO!
```
Comment (dnadales) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:98]: We could address this in this PR or as a follow-up issue.

File: ouroboros-consensus-shelley-test/src/Test/ThreadNet/TxGen/Shelley.hs [hunk 2]
```diff
@@ -63,12 +64,9 @@ instance HashAlgorithm h => TxGen (ShelleyBlock (TPraos (MockCrypto h)) (MockShe
...
 
-      | otherwise               =
-      if True
-        then pure []
-        else do
-          n <- choose (0, 20)
-          go [] n $ applyChainTick lcfg curSlotNo lst
+      | otherwise               = do
+      n <- choose (0, 20)
+      go [] n $ applyLedgerTablesDiffsTicked lst $ applyChainTick lcfg curSlotNo $ forgetLedgerTables lst
```
Comment (dnadales) [inline ouroboros-consensus-shelley-test/src/Test/ThreadNet/TxGen/Shelley.hs:69]: We can maybe wrap this one.

**Review by dnadales — APPROVED** [review]
There are a couple of rough edges we could smooth out before merging this, but we could also do this afterwards to unblock us (and reduce the scope of this PR).

File: ouroboros-consensus-shelley-test/src/Test/Consensus/Shelley/Examples.hs [hunk 3]
```diff
@@ -230,13 +264,33 @@ fromShelleyLedgerExamplesPraos ShelleyLedgerExamples {
...
+                  -- transaction output) in the example provided by
+                  -- cardano-ledger to make sure that we test the serialization
+                  -- of ledger tables with at least one non-trivial example.
+            xs -> xs
 
+        exampleTxOuts :: [LC.TxOut era]
+        exampleTxOuts =
+          case toList (sleTx ^. LC.bodyTxL ^. LC.outputsTxBodyL) of
+            [] -> error "No transaction outputs were provided to construct the ledger tables"
+            xs -> xs
```
Comment (jorisdral) [inline ouroboros-consensus-shelley-test/src/Test/Consensus/Shelley/Examples.hs:293]: This is duplicated code -- maybe factor out a `mkLedgerTables` function from `fromShelleyLedgerExamplesPraos` and `fromShelleyLedgerExamples`?

**Review by jorisdral — APPROVED** [review]
> There are a couple of rough edges we could smooth out before merging this, but we could also do this afterwards to unblock us (and reduce the scope of this PR).

Agreed

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs (hunk 1 above)
Comment (Jasagredo) [reply ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:89]: Done [#4473](https://github.com/input-output-hk/ouroboros-network/issues/4473)

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+  , tsDestLedgerConfig :: LedgerConfig dest
+  , tsSrcLedgerState   :: LedgerState src EmptyMK
+  , tsEpochNo          :: EpochNo
+}
+
+deriving instance ( Show (LedgerConfig src)
+                  , Show (LedgerConfig dest)
+                  , Show (LedgerState src EmptyMK)) => Show (TestSetup src dest)
+
+-- TODO: Useful to merge some of these instances?
```
[the remark this answers is not in the document]
Comment (Jasagredo) [reply ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:293]: I think we can ignore it for now. Maybe they cannot even be merged

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+  -- ^ Property label
+  -> RequiringBoth
+        WrapLedgerConfig
+        TranslateLedgerState
+        srcBlk
+        dstBlk
+  -> (LedgerState srcBlk EmptyMK -> LedgerState dstBlk DiffMK -> Bool)
+  -> (LedgerState srcBlk EmptyMK -> Property -> Property)
+  -- ^ Coverage testing function (TODO: it can be abused)
+  -> TestTree
```
[the remark this answers is not in the document]
Comment (Jasagredo) [reply ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs]: It is fine, we are the only consumers of this

File: ouroboros-consensus-byron-test/src/Test/Consensus/Byron/Examples.hs [hunk 6]
```diff
@@ -184,22 +189,28 @@ emptyLedgerState = ByronLedgerState {
...
+    . applyLedgerTablesDiffsTicked emptyLedgerState
     . applyChainTick ledgerConfig (SlotNo 0)
+    . forgetLedgerTables
     $ emptyLedgerState
 
-exampleLedgerState :: LedgerState ByronBlock
+exampleLedgerState :: LedgerState ByronBlock ValuesMK
 exampleLedgerState =
-      reapplyLedgerBlock ledgerConfig exampleBlock
+      applyLedgerTablesDiffs emptyLedgerState
```
[the remark this answers is not in the document]
Comment (Jasagredo) [reply ouroboros-consensus-byron-test/src/Test/Consensus/Byron/Examples.hs:196]: I don't think it makes much of a difference, and I think it was already applying on top of the empty ledger state, not on top of the lsAfterEBB

File: ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/Examples.hs [hunk 7]
```diff
@@ -117,21 +126,22 @@ instance Inject SomeResult where
...
+      , exampleHeader           = inj (Proxy @Header)                        exampleHeader
+      , exampleSerialisedHeader = inj (Proxy @SerialisedHeader)              exampleSerialisedHeader
+      , exampleHeaderHash       = inj (Proxy @WrapHeaderHash)                exampleHeaderHash
+      , exampleGenTx            = inj (Proxy @GenTx)                         exampleGenTx
+      , exampleGenTxId          = inj (Proxy @WrapGenTxId)                   exampleGenTxId
+      , exampleApplyTxErr       = inj (Proxy @WrapApplyTxErr)                exampleApplyTxErr
+      , exampleQuery            = inj (Proxy @(SomeSecond BlockQuery))       exampleQuery
+      , exampleResult           = inj (Proxy @SomeResult)                    exampleResult
+      , exampleAnnTip           = inj (Proxy @AnnTip)                        exampleAnnTip
+      , exampleLedgerState      = inj (Proxy @(Flip LedgerState EmptyMK))    exampleLedgerState
```
[the remark this answers is not in the document]
Comment (Jasagredo) [reply ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/Examples.hs:130]: https://github.com/input-output-hk/ouroboros-network/issues/4477

**Pull request merged**